packages/keystatic/src/reader/generic.ts

```diff
@@ -0,0 +1,445 @@
+  parent: string,
+  fsReader: MinimalFs
+): Promise<{ entry: DirEntry; name: string }[]> {
+  return (
+    await Promise.all(
+      (await fsReader.readdir(parent)).map(async dirent => {
+        const name = `${parent}${dirent.name}`;
+        const entry = { entry: dirent, name };
+        if (dirent.kind === 'directory') {
+          return [entry, ...(await getAllEntries(`${name}/`, fsReader))];
+        }
+        return entry;
+      })
+    )
+  ).flat();
+}
+
+const listCollection = cache(async function listCollection(
+  collectionPath: string,
+  glob: Glob,
+  formatInfo: FormatInfo,
+  extension: string,
+  fsReader: MinimalFs
+) {
+  const entries: { entry: DirEntry; name: string }[] =
+    glob === '*'
+      ? (await fsReader.readdir(collectionPath)).map(entry => ({
+          entry,
+          name: entry.name,
+        }))
+      : (await getAllEntries(`${collectionPath}/`, fsReader)).map(x => ({
+          entry: x.entry,
+          name: x.name.slice(collectionPath.length + 1),
+        }));
+
+  return (
+    await Promise.all(
+      entries.map(async x => {
+        if (formatInfo.dataLocation === 'index') {
+          if (x.entry.kind !== 'directory') return [];
+          if (
+            !(await fsReader.fileExists(
+              getEntryDataFilepath(`${collectionPath}/${x.name}`, formatInfo)
+            ))
+          ) {
+            return [];
+          }
+          return [x.name];
+        } else {
+          if (x.entry.kind !== 'file' || !x.name.endsWith(extension)) {
+            return [];
+          }
+          return [x.name.slice(0, -extension.length)];
+        }
+      })
+    )
+  ).flat();
+});
+
+export function collectionReader(
+  collection: string,
+  config: Config,
+  fsReader: MinimalFs
+): CollectionReader<any, any> {
+  const formatInfo = getCollectionFormat(config, collection);
+  const collectionPath = getCollectionPath(config, collection);
+  const collectionConfig = config.collections![collection];
+  const schema = fields.object(collectionConfig.schema);
+  const glob = getSlugGlobForCollection(config, collection);
+  const extension = getDataFileExtension(formatInfo);
+
+  const read: CollectionReader<any, any>['read'] = (slug, ...args) =>
+    readItem(
+      schema,
+      formatInfo,
+      getCollectionItemPath(config, collection, slug),
+      args[0]?.resolveLinkedFiles,
+      `"${slug}" in collection "${collection}"`,
+      fsReader,
+      slug,
+      collectionConfig.slugField,
+      glob
+    );
+
+  const list = () =>
+    listCollection(collectionPath, glob, formatInfo, extension, fsReader);
+
+  return {
+    read,
+    readOrThrow: async (...args) => {
+      const entry = await (read as any)(...args);
+      if (entry === null) {
+        throw new Error(
+          `Entry "${args[0]}" not found in collection "${collection}"`
+        );
+      }
+      return entry;
+    },
+    // TODO: this could drop the fs.stat call that list does for each item
```

Comment:
This doesn't seem relevant to the code below?

Reply:
It is, `list` does checks that aren't really necessary since we immediately read the item in `all`